craft_providers/pro.py

```diff
@@ -0,0 +1,91 @@
+#
+# Copyright 2021-2024 Canonical Ltd.
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License version 3 as published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public License
+# along with this program; if not, write to the Free Software Foundation,
+# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+"""Pro-related module."""
+
+import json
+import logging
+
+import requests
+
+from craft_providers.errors import MachineTokenError
+
+logger = logging.getLogger(__name__)
+
+
+class ProToken:
+    """Class for retrieving and processing pro tokens."""
+
+    @staticmethod
+    def retrieve_pro_host_token() -> str:
+        """Get the machine token from the pro host."""
+        try:
+            with open("/var/lib/ubuntu-advantage/private/machine-token.json") as fd:
+                content = json.load(fd)
+                machine_token = content.get("machineToken", "")
+                if not machine_token:
+                    raise MachineTokenError("No token in machine token file.")
+                return machine_token
+        except FileNotFoundError:
+            raise MachineTokenError("Machine token file does not exist.")
+        except PermissionError:
+            raise MachineTokenError(
+                "Machine token file is not accessible. Make sure you are "
+                "running with root access."
+            )
+
+    @staticmethod
+    def request_pro_guest_token() -> str:
+        """Request a guest token from contracts API."""
+        machine_token = ProToken.retrieve_pro_host_token()
+
+        try:
+            base_url = "https://contracts.canonical.com"
+            endpoint = "/v1/guest/token"
+            response = requests.get(
+                base_url + endpoint,
+                headers={"Authorization": f"Bearer {machine_token}"},
+                timeout=15,
+            )
+            if response.status_code != 200:
+                # fallback mechanism
```

Comment:
what is this fallback mechanism here? I don't remember us talking about it, shouldn't it fail if we can't get a guest token?

Reply:
This is a temporary fallback mechanism because both the contracts API and the Pro backend's corresponding endpoint are not available yet. Once those are available we can remove this fallback.
For more info see https://warthogs.atlassian.net/browse/ROCKS-1460.